Add `kotlin.serialization` implementation of `Serializer`

hiddewie opened the pull request:
(Followup from #121 which had to be closed because I deleted my Github fork.) 

This PR adds support for Kotlin serialization, see issue #13.

This PR includes #120 (this is required in order to use the stable version of `kotlin.serialization`).

In summary:
- Adds a class `KotlinSerializer` which implements Axon `Serializer`.
- Configurable `RevisionResolver`, `Converter`, and `Json`, with the same defaults as the Jackson serializer implementation. 
- Because `Json` is configurable, applications can define how JSON should be serialized based on business requirements.
- Adds a DSL function `kotlinSerializer` to create a `KotlinSerializer` with configuration options.
- Useful errors when the `.serializer()` method cannot be found on a companion object of a to-serialize object.
- `Class -> KSerializer` is cached for performance and to avoid too much reflection.
- Simple test to show that the implementation works.

Ref https://kotlinlang.org/docs/serialization.html with basic overview.
Ref https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/serialization-guide.md with serialization guide.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+                        typeForClass(type)
+                    )
+
+                else ->
```
[]I would've assumed the `Serializer` could default to `byte[]` instead of throwing an exception.
Doing so will have this serializer follow the same paradigm as the `JacksonSerializer` and `XStreamSerializer`.

Several spots in the framework assume that the `expectedRepresentation` is `byte[]`.
So enabling this option up would make this serializer more valuable throughout the framework.

smcvb replied:
Pretty confident this could this will require introducing the `registerConverters` method (as seen on line 128 of the `JacksonSerializer` too).
This method ensures Json specific `ContentTypeConverters` are added to the set-up. 

Furthermore, I assume this will warrant introducing content type converters going from `JsonElement` to `byte[]`, and vice versa.

hiddewie replied:
I totally forgot about the (extra) converters. I will add some defaults to convert between byte arrays <-> Strings and JSON Nodes :+1:

hiddewie replied:
I made the `JsonElement` type the exception, just like `JsonNode` in the Jackson serializer. All other target types are serialized to String, and then converted with the chaining converter.

By default String, Byte array, Json Element and InputStream already work through service discovery, so that is fine.

If any other converters are needed I can add them.

smcvb replied:
Thanks for validating @hiddewie :+1: 
Adding tests to validate this works (as in, the conversion between those types through this serializer) are a massive plus.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    override fun <T> canSerializeTo(expectedRepresentation: Class<T>): Boolean =
+        expectedRepresentation == String::class.java ||
```
Once the `serialize` method is adjusted, this method should include the `Converter` too.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+
+            val serializer: KSerializer<T> = serializedObject.serializer()
+            return when {
+                serializedObject.contentType.isAssignableFrom(String::class.java) ->
```
I am missing the `UnknownSerializedType` validation here. This is required to allow the `Serializer` to deal with unknown objects (like old/new events the application isn't familiar with) without breaking down the (distributed) setup.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+                serializedObject.contentType.isAssignableFrom(JsonElement::class.java) ->
+                    json.decodeFromJsonElement(serializer, serializedObject.data as JsonElement)
+
+                else ->
```
Just as with the `serialize` method, I miss the `byte[]` option here.

hiddewie replied:
:+1:

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+            // Class<T>: T must be non-null
+            val kClass = (this as Class<Any>).kotlin
+
+            val companion = kClass.companionObject
```
I assume that the hard requirement on the `@Serializable` annotation makes it pretty tough to use this serializer for Axon objects, like the `TrackingToken`.
I've skimmed the Kotlin Serializer documentation somewhat but didn't spot a solution for this just yet.
Do you, perchance, have a solution in mind for this, @hiddewie?

hiddewie replied:
This is the toughest question :)

As far as I can see, there are two solutions:
1. Annotate the serializable types in Axon core with `@Serializable`. This adds the compile time requirement of having the Kotlin compiler plugin dependency in the build process. That may or may not be OK. Although, there are also Jackson annotations in the source code already. In runtime there is no Kotlin requirement.
2. Create handwritten serializers for every class that needs to be serializable. This is quite some (cumbersome, error prone) work, depending on the number of serializable classes in the Axon core. The serializer code can be maintained in this repository.

To illustrate option 2, I added a small example and custom serializer with test.

The third option is abandoning this approach because the Kotlin serialization is not wanted in the core framework, and it is not worth the effort to maintain the custom serializers.

hiddewie replied:
(After thinking about option 1 a bit more: I have not tested adding `@Serializable` on a Java class and then running the Kotlin compiler plugin on it. Theoretically it should work but in practice I don't know)

hiddewie replied:
I tried a few things to test how the interoperability of Java classes with Kotlin annotations, and the Kotlin annotation processors work together.

What I did:
- Complie Axon framewok locally (4.6.0-SNAPHSOT)
- Add the Maven Kotlin compiler plugin which would compile the Kotlin annotations into `serializer` functions, see instructions here https://github.com/Kotlin/kotlinx.serialization#maven
- Add `@kotlinx.serialization.Serializable` to ConfigToken

What I saw:
- The annotation gets added (just like any other annotation) in the classfile
- No serializer static method is generated.

What I did:
- Copy the Java class and copy it to this repository, with the .java extension. 
- Let the Kotlin compiler, with the Kotlin serialization compiler plugin comple the .java class as well

What I saw:
- The annotation gets added in the classfile
- Still no serializer static method  is generated.

What I did:
- Let IntelliJ convert the .java file to a .kt Kotlin file (very ugly)

What I saw:
- The annotation gets added in the classfile
- A serializer method is generated.

So my conclusion is that the Kotlin serialization compiler plugin only works for Kotlin source files, and that the `.serializer()` static method cannot be generated for Java source files. That is very unfortunate, I might open an issue with the Kotlin Serialization team to ask if that is supposed to work.

sandjelkovic replied:
Either way, we can not assume that every class that comes here will have a serializer attached to it (nor be a Kotlin class)

If there is no serializer defined, or just based on types, we can fall back on a registry that we can keep as a map of Type (Class, Class name, etc) and a Serializer. That map could be configurable so someone might add support for classes without `@Serializable` like Java Dates, _and_ we can pre-populate it with AF serializers configuration. Or keep the AF serializers in a separate config, but that's a finer detail.

[This](https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#passing-a-serializer-manually) and 
[this](https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#passing-a-serializer-manually) is what I mean, as we need to pass the serializer explicitly in the very broad generic setting anyway, we might as well keep it somewhere in a config map, similar as to how this cache is doing but is just caching _all_ serializers.

smcvb commented on kotlin/src/test/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializerTest.kt:
```diff
@@ -0,0 +1,95 @@
+import kotlin.test.assertNull
+
+class KotlinSerializerTest {
+
```
I am missing test cases that validate if the `KotlinSerializer` can be used for Axon objects, like the `Message`, `TrackingToken`, and `SagaEntry`.

Assuming that's because those objects aren't annotated with the `@Serializable` annotation.
However, the `Serializer` in Axon must de-/serialize those objects for it to be usable.

hiddewie replied:
Indeed. See https://github.com/AxonFramework/extension-kotlin/pull/124/files#r645048334 for the comment about serializing internal classes.

I added one test case for the `ConfigToken`, and every internal class that should be serializable can be added in that way, if there is a serializer.

smcvb replied:
Gotcha. To make this `Serializer` implementation workable for Axon Framework entirely, these tests will be a requirement too.

Making these can wait until we've concluded the conversation on the `Serializer` implementation though.
On the subject of using the `@Serializable` annotation on Axon components, and if that even works for Java classes, that is.

smcvb reviewed (changes requested):
Bunch of comments on here. The biggest one is that it does not contain how this `Serializer` should be used for Axon's internal objects. For this to be useable within Axon, I am afraid it must support Axon's objects like the `Message`, `TrackingToken`, and `SagaEntry`.

hiddewie commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+ * @see kotlinSerializer
+ */
+class KotlinSerializer(
+    private val revisionResolver: RevisionResolver,
```
The defaults are provided at one place, in the parameters constructed in the DSL function. That is also the place where the defaults can be modified by adding an initializer function.

I can also add them here as default arguments.

smcvb replied:
That's more in line with all the other infrastructure components Axon Framework, and the Extensions provide.
I don't intend to be religious about this, though, as I pointed out under the `KotlinSerializerConfiguration` comment.

hiddewie replied:
:+1: Done

hiddewie commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+            SimpleSerializedType(type.name, revisionResolver.revisionOf(type))
+        }
+
+    private fun revisionOf(type: Class<*>): String? =
```
Yeah, I inlined it :)

hiddewie commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,205 @@
+ * @see KotlinSerializer
+ * @see kotlinSerializer
+ */
+class KotlinSerializerConfiguration {
```
These properties are all `var`, so mutable. This makes a clean DSL possible, without invoking the constructor directly. But the DSL is totally optional and I could remove it.

smcvb replied:
Honestly, I wouldn't know what's more "Kotlin-like". 

If I'd look at the other Axon components, the settings are done through the Builder of a piece of infrastructure.
We've quite recently allowed Builder configuration directly through Axon's overall `Configurer` (for the `PooledStreamingEventProcessor`, to be exact), which we feel might be the way forward.

hiddewie replied:
I removed this Configuration class, and used the constructor with default values.

hiddewie commented:
Thank you for the great comments @smcvb. The discussion in https://github.com/AxonFramework/extension-kotlin/pull/124/files#r645048334 remains unresolved which will require more changes whatever solution is chosen.

hiddewie commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/serializer/ConfigTokenSerializer.kt:
```diff
@@ -0,0 +1,49 @@
+
+@OptIn(ExperimentalSerializationApi::class)
+@Serializer(forClass = ConfigToken::class)
+class ConfigTokenSerializer : KSerializer<ConfigToken> {
```
This is the custom serializer implemented manually for the internal Axon type. These type of implementations are needed if the `@Serializable` annotation is not used in combination with the compiler plugin.

smcvb replied:
I'd assume this should be constructed for every concrete implementation of serializable Axon objects in that case, correct?

This would at least enable moving further with this serializer, so that's good.

hiddewie replied:
I just saw https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#deriving-external-serializer-for-another-kotlin-class-experimental for the first time. This would allow us to skip the 'manual' implementation of the serializer.

Unfortunately I can't yet get it to work, I run into this bug  in the Kotlin complier plugin (https://github.com/Kotlin/kotlinx.serialization/issues/1680).

hiddewie replied:
(https://github.com/AxonFramework/extension-kotlin/pull/124#issuecomment-1100655939) I changed the fully-custom example `ConfigTokenSerializer` into a serializer using a `ConfigTokenSurrogate`. Its slightly better maintainable, but still a LOT of work to do (and maintain) for every Axon class.

The work means, for every Axon class:
- Create a Kotlin class with `@Serializable`
- Put all the 
- Manually implement a serializer for that class, using the SerialDescriptor generated by the surrogate.
- During serialization, convert from the real value to a surrogate and serialize it.
- During deserialization, deserialize the surrogate and convert from the surrogate to the value of the real class.

The serializers for surrogates are very simple and similar (only a few lines of code). However the surrogates themselves need to contain exactly the public data fields of the class they are a surrogate for. This will cause bugs if a field is forgotten in the surrogate.

smcvb commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/serializer/ConfigTokenSerializer.kt:
```diff
@@ -0,0 +1,49 @@
+import kotlinx.serialization.encoding.encodeStructure
+import org.axonframework.eventhandling.tokenstore.ConfigToken
+
+@OptIn(ExperimentalSerializationApi::class)
```
Would you be able to elaborate a little on the `ExperimentalSerializationApi` you're "opting in" here?

hiddewie replied:
The `@Serializer` annotation below is marked as experimental. It might change its implementation slightly between versions of Kotlin Serialization. (https://kotlin.github.io/kotlinx.serialization/kotlinx-serialization-core/kotlinx-serialization-core/kotlinx.serialization/-experimental-serialization-api/index.html)

We might wait to merge this feature until the used serialization API is marked fully stable, although the Axon `extension-kotlin` library  is also marked as experimental itself.

hiddewie replied:
Yes exactly. The following is stable:
```
@Serializable
class XXX
```

But the following needs experimental opt in:
```
@Serializer(forClass = QQQ::class)
class XXX
```

In particular doing this for Java classes gives compiler errors, and it might simply not be supported (see https://github.com/Kotlin/kotlinx.serialization/issues/1687)

smcvb reviewed (commented):
Sorry for returning so late here. Times are rather busy, and some things sadly got precedence. 
On any note, I think it is paramount we first tackle the Axon Serializable components predicament, specifically on whether that annotation would be an option. From my end, I'll chat with the team if we feel it has a time and place inside the core.

hiddewie commented:
No problem :)

I investigated the compilation of the `@Serializable` annotation in Axon Java source code. That does not look like it will work.

Then I found https://github.com/Kotlin/kotlinx.serialization/blob/f673bc2/docs/serializers.md#deriving-external-serializer-for-another-kotlin-class-experimental, which could mean that we create singleton objects marked as `@Serializer(forClass = ...)` that would auto-generate the needed serializers within the compilation process of this library. Unfortunately it doesn't work yet: the compilation crashes.

smcvb commented:
FYI, @hiddewie, I've adjusted the milestone to 0.3.0 instead of 0.2.0. 

The intent for this is the desire to release 0.2.0 of the Kotlin extension. 
We felt we couldn't wait much longer until this release.

This, by the way, doesn't mean a 0.3.0 release couldn't be around the corner, which might contain this serializer. ;-)

hiddewie commented:
Sorry for the long wait, I didn't forget the PR but it took some time until I could dive into it again!

I updated this branch to include all changes in `master`.

I updated the Kotlin serialization version to 1.3.2, but this makes no changes to everything I have observed so far.

In particular I experimented with a Surrogate serializer (https://github.com/AxonFramework/extension-kotlin/pull/124#discussion_r817801799). This works, but still requires a lot of maintenance work to 'wrap' each Axon internal class with a surrogate class in this library, while copying the public fields manually into the surrogate class for serialization. Performance-wise this is also not ideal. 

https://github.com/AxonFramework/extension-kotlin/pull/124/files#diff-cc9e7866a7929cd55740daaa367352b89517dc56787fa6eb48f5c2fcb783b1bbR12-R31

I tested the automatically-generated serializer for external classes functionality again, but this still gives compiler errors: 
```
@Serializer(forClass = ConfigToken::class)
object ConfigTokenSerializer 
```
If I manually type the simple class like `ConfigToken` using Kotlin, it does work. So there might again be a Kotlin <-> Java interop problem here, causing errors in the Kotlin compiler.

hiddewie commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/serialization/KotlinSerializer.kt:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    override fun <T> canSerializeTo(expectedRepresentation: Class<T>): Boolean =
+        expectedRepresentation == JsonElement::class.java ||
```
Yes definitely :+1: Done

smcvb commented:
Removing the milestone as this is still under discussion, while the release is intended to come soon. Will add this to the following release milestone for further discussions.

smcvb commented:
Marking this pull request as obsolete through the introduction of #338.
